server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java

```diff
@@ -417,14 +428,53 @@ private EmbeddedZooKeeper startZooKeeper() throws Exception {
             System.setProperty("zookeeper.authProvider.1", SASLAuthenticationProvider.class.getName());
 
             // Set the client port, which is unused.
-            zkProps.setProperty("clientPort", String.valueOf(cfg.serverAddress().clientPort()));
+            zkProps.setProperty("clientPort", String.valueOf(cfg.serverConfig().clientPort()));
 
+            final Map<Integer, ZooKeeperServerConfig> servers = cfg.servers();
             // Add replicas.
-            cfg.servers().forEach((id, addr) -> {
+            boolean hasGroupId = false;
+            for (Entry<Integer, ZooKeeperServerConfig> entry : servers.entrySet()) {
+                final ZooKeeperServerConfig serverConfig = entry.getValue();
                 zkProps.setProperty(
-                        "server." + id,
-                        addr.host() + ':' + addr.quorumPort() + ':' + addr.electionPort() + ":participant");
-            });
+                        "server." + entry.getKey(),
+                        serverConfig.host() + ':' + serverConfig.quorumPort() + ':' +
+                        serverConfig.electionPort() + ":participant");
+
+                if (!hasGroupId && serverConfig.groupId() != null) {
+                    hasGroupId = true;
+                }
+            }
+
+            // Add groups if exists
+            if (hasGroupId) {
+                final ImmutableMultimap.Builder<Integer, Integer> groupBuilder = ImmutableMultimap.builder();
+                boolean isHierarchical = true;
+                for (Entry<Integer, ZooKeeperServerConfig> entry : servers.entrySet()) {
+                    final Integer groupId = entry.getValue().groupId();
+                    if (groupId == null) {
+                        isHierarchical = false;
+                        final List<ZooKeeperServerConfig> noGroupIds =
+                                servers.values().stream()
+                                       .filter(serverConfig -> serverConfig.groupId() == null)
+                                       .collect(toImmutableList());
+                        logger.warn("Hierarchical quorums are disabled. 'groupId' are missing in {}",
+                                    noGroupIds);
+                        break;
+                    } else {
+                        groupBuilder.put(groupId, entry.getKey());
+                    }
+                }
+                if (isHierarchical) {
+                    groupBuilder.build().asMap().forEach((groupId, serverIds) -> {
+                        // group.1=1:2:3
+                        zkProps.setProperty("group." + groupId, colonJoiner.join(serverIds));
+                    });
+                    servers.forEach((serverId, serverConfig) -> {
+                        // weight.1=1
```

Comment:
How about prepending `e.g.` to the examples in the comments?

Reply:
Nice suggestion!